.azure-pipelines-steps.yml

```diff
@@ -22,6 +22,12 @@ steps:
   - script: node scripts/remove-postinstall
     displayName: 'Remove postinstall script'
 
+  - task: CacheBeta@0
+    inputs:
+      key: yarn | $(Agent.OS) | yarn.lock
+      path: $(YARN_CACHE_FOLDER)
```

Comment:
Btw, there is a caching feature coming in the next few weeks (tarring cache contents) that will make it easier/possible to cache `node_modules`. This generally results in better caching performance compared to caching the Yarn (or npm) global cache folder. 

Definitely curious to see what improvements you get here. Keep in mind the first "save cache" operation within an org typically takes longer than future save operations (so don't be surprised if your first run takes longer than usual).

Reply:
Awesome, that sounds great! That's what we cache on circus as well. I'll set up a reminder for myself to check back in a couple of weeks 🙂

Reply:
Is there an issue I could track for this work? I tried searching on https://github.com/microsoft/azure-pipelines-yaml (dunno if it's even the correct repo) without finding anything

Reply:
Hey @SimenB - see https://github.com/microsoft/azure-pipelines-tasks/issues/10925

Net-net: once your hosted agent's version (which is reported in the `Initialize job` step) shows as 2.157.0 or later, you can try caching the `node_modules` folder  after setting the `AZP_CACHING_TAR` variable to `true`.

Reply:
Awesome, thanks! Seems we're `2.155` for now, so I'll check back in a few days 🙂

Reply:
@willsmythe I see we're getting the `2.160` agents now. Do you still recommend to change anything? The docs still state to cache the yarn cache folder: https://docs.microsoft.com/en-us/azure/devops/pipelines/caching/?view=azure-devops#nodejsyarn

Reply:
@SimenB - yah we should try caching `node_modules` using the released version of the Cache task (V2). I'll test out a few options.

Reply:
Here are a few tests. Note: I wouldn't say this is totally scientific because, as we have seen, there is some variability in terms of build time from one run to the next (even without any changes):

| Option            | Install dependencies (Linux) | Run tests (Windows) | Total time |
|-------------------|:----------------------------:|:----------:|:---:|
| A. [Yarn global](https://dev.azure.com/willsmythe/jest/_build/results?buildId=1364&view=results) | 1m 7s  | 18m 22s (84%) | 21m 57s |
| B. [`node_modules`](https://dev.azure.com/willsmythe/jest/_build/results?buildId=1363&view=results) | 1m 52s  | 18m 50s (80%) | 23m 51s |
| C. [`node_modules` and Yarn global](https://dev.azure.com/willsmythe/jest/_build/results?buildId=1361&view=results) | 37s | 18m 42s (86%) | 21m 41s |

Net-net: I don't think Jest will see huge performance improvements by switching to caching just `node_modules` versus Yarn's global cache directory. The added time to restore/save the extra bytes and thousands of files seems to offset any gains.

It is interesting that caching both (option C) resulted in better "install dependency" times, but just caching node_modules resulted in significantly worse times.

I experimented some with Yarn's offline cache, but didn't see any significant improvements.

The third column highlights where most of the overall time is spent (**running tests**). If you look at the [Tests results](https://dev.azure.com/willsmythe/jest/_build/results?buildId=1372&view=ms.vss-test-web.build-test-results-tab) for a run and filter to include **Passing** tests and order by **Duration**, you can see there are numerous tests (in all platforms) that take over 5s each. This obviously adds up over time.

![image](https://user-images.githubusercontent.com/2503052/69062316-bf041580-09e8-11ea-8db1-021204ff4707.png)

Based on this, **the best way to speed up Jest test runs is probably to split Jest test runs across more pipeline jobs**,

Reply:
Ah, ok. We'll leave it alone then. Thank you so much for diving into this @willsmythe!